homeassistant/components/mazda/__init__.py

```diff
@@ -0,0 +1,161 @@
+"""The Mazda Connected Services integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+import async_timeout
+from pymazda import (
+    Client as MazdaAPI,
+    MazdaAccountLockedException,
+    MazdaAPIEncryptionException,
+    MazdaAuthenticationException,
+    MazdaException,
+    MazdaTokenExpiredException,
+)
+
+from homeassistant.config_entries import SOURCE_REAUTH, ConfigEntry
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD, CONF_REGION
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import DATA_CLIENT, DATA_COORDINATOR, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Mazda Connected Services component."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Mazda Connected Services from a config entry."""
+    email = entry.data[CONF_EMAIL]
+    password = entry.data[CONF_PASSWORD]
+    region = entry.data[CONF_REGION]
+
+    websession = aiohttp_client.async_get_clientsession(hass)
+    mazda_client = MazdaAPI(email, password, region, websession)
+
+    try:
+        await mazda_client.validate_credentials()
+    except MazdaAuthenticationException:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_REAUTH},
+                data=entry.data,
+            )
+        )
+        return False
+    except (
+        MazdaException,
+        MazdaAccountLockedException,
+        MazdaTokenExpiredException,
+        MazdaAPIEncryptionException,
+    ) as ex:
+        _LOGGER.error("Error occurred during Mazda login request: %s", ex)
+        raise ConfigEntryNotReady from ex
+
+    async def async_update_data():
+        """Fetch data from Mazda API."""
+        try:
+            async with async_timeout.timeout(10):
+                vehicles = await mazda_client.get_vehicles()
+                for vehicle in vehicles:
+                    status = await mazda_client.get_vehicle_status(vehicle["id"])
```

Comment:
You're doing an await for each vehicle but it's inside a timeout block for 10 seconds. The odds of someone hitting this gets higher if they have more cars. We should do a timeout per request. Also, not sure about API rate limits, but if possible, try to get all vehicle statuses at once. 

If you don't want to ddos the API if someone has 100 cars but still provide maximum parallelism, you can use `gather_with_concurrency` from `homeassistant.util.async_`.

Something like this:

```python
# Not tested, from memory :)

async def with_timeout(task):
    async with async_timeout.timeout(10):
        return await task

async with async_timeout.timeout(10):
    vehicles = await mazda_client.get_vehicles()

statuses = await gather_with_concurrency(5, [
    with_timeout(mazda_client.get_vehicle_status(vehicle["id"]))
    for vehicle in vehicles
])

for vehicle, status in zip(vehicles, statuses):
    vehicle['status'] = status
```

Reply:
That's a great point, I didn't consider this. I updated the code to use `gather_with_concurrency`.